File: ios/XLibReader/Reader/ByteOffsetMap.swift
import Foundation

struct ByteOffsetMap: Sendable {
    private let text: String
    private let encoding: TextEncoding
    private let characterAnchors: [Int]
    private let byteAnchors: [Int]

    init(text: String, encoding: TextEncoding, stride: Int = 128) throws {
        self.text = text
        self.encoding = encoding
        let safeStride = max(16, stride)
        let source = text as NSString
        var characters = [0]
        var bytes = [0]
        var previousCharacter = 0
        var byteOffset = 0

        while previousCharacter < source.length {
            var nextCharacter = min(source.length, previousCharacter + safeStride)
            if nextCharacter < source.length,
               Self.isLowSurrogate(source.character(at: nextCharacter)),
               nextCharacter > previousCharacter {
                nextCharacter -= 1
            }
            if nextCharacter <= previousCharacter {
                nextCharacter = min(source.length, previousCharacter + 1)
            }
            guard let length = encoding.encodedByteCount(
                of: source.substring(with: NSRange(location: previousCharacter, length: nextCharacter - previousCharacter))
            ) else {
                throw ReaderCoreError.invalidByteMap
            }
            byteOffset += length
            characters.append(nextCharacter)
            bytes.append(byteOffset)
            previousCharacter = nextCharacter
        }

        characterAnchors = characters
        byteAnchors = bytes
    }

    var utf16Length: Int { (text as NSString).length }
    var totalBytes: Int { byteAnchors.last ?? 0 }

    func byteOffset(forUTF16Index index: Int) throws -> Int {
        let source = text as NSString
        var safeIndex = max(0, min(index, source.length))
        if safeIndex > 0,
           safeIndex < source.length,
           Self.isLowSurrogate(source.character(at: safeIndex)) {
            safeIndex -= 1
        }
        let anchor = Self.floorIndex(in: characterAnchors, target: safeIndex)
        let range = NSRange(location: characterAnchors[anchor], length: safeIndex - characterAnchors[anchor])
        guard let delta = encoding.encodedByteCount(of: source.substring(with: range)) else {
            throw ReaderCoreError.invalidByteMap
        }
        return byteAnchors[anchor] + delta
    }

    func utf16Index(forByteOffset offset: Int) throws -> Int {
        let target = max(0, min(offset, totalBytes))
        let anchor = Self.floorIndex(in: byteAnchors, target: target)
        let source = text as NSString
        var characterIndex = characterAnchors[anchor]
        var currentByte = byteAnchors[anchor]

        while characterIndex < source.length {
            let nextCharacter = nextUTF16Boundary(in: source, after: characterIndex)
            let value = source.substring(
                with: NSRange(location: characterIndex, length: nextCharacter - characterIndex)
            )
            guard let length = encoding.encodedByteCount(of: value) else {
                throw ReaderCoreError.invalidByteMap
            }
            if currentByte + length > target { break }
            currentByte += length
            characterIndex = nextCharacter
        }
        return characterIndex
    }

    /// Returns the first complete UTF-16 character boundary whose encoded byte
    /// offset is greater than or equal to `offset`.
    func utf16IndexAtOrAfter(byteOffset offset: Int) throws -> Int {
        let target = max(0, min(offset, totalBytes))
        let floor = try utf16Index(forByteOffset: target)
        let floorByte = try byteOffset(forUTF16Index: floor)
        guard floorByte < target else { return floor }
        return nextUTF16Boundary(in: text as NSString, after: floor)
    }

    private func nextUTF16Boundary(in source: NSString, after index: Int) -> Int {
        guard index + 1 < source.length else { return source.length }
        let current = source.character(at: index)
        let next = source.character(at: index + 1)
        if Self.isHighSurrogate(current), Self.isLowSurrogate(next) {
            return index + 2
        }
        return index + 1
    }

    private static func floorIndex(in values: [Int], target: Int) -> Int {
        var low = 0
        var high = values.count - 1
        while low <= high {
            let middle = (low + high) / 2
            if values[middle] <= target {
                low = middle + 1
            } else {
                high = middle - 1
            }
        }
        return max(0, high)
    }

    private static func isHighSurrogate(_ value: unichar) -> Bool {
        (0xD800...0xDBFF).contains(value)
    }

    private static func isLowSurrogate(_ value: unichar) -> Bool {
        (0xDC00...0xDFFF).contains(value)
    }
}

File: apps/android/app/src/main/java/com/xlib/txtreader/ByteOffsetMap.java
package com.xlib.txtreader;

import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.List;

/** Sparse, bidirectional mapping between Java UTF-16 indices and source bytes. */
final class ByteOffsetMap {
    private static final int DEFAULT_STRIDE = 128;

    private final String text;
    private final Charset charset;
    private final int[] charAnchors;
    private final int[] byteAnchors;

    static ByteOffsetMap create(String text, Charset charset) {
        return new ByteOffsetMap(text == null ? "" : text, charset, DEFAULT_STRIDE);
    }

    ByteOffsetMap(String text, Charset charset, int stride) {
        this.text = text == null ? "" : text;
        this.charset = charset;
        int safeStride = Math.max(16, stride);
        List<Integer> chars = new ArrayList<>();
        List<Integer> bytes = new ArrayList<>();
        chars.add(0);
        bytes.add(0);

        int previousChar = 0;
        int byteOffset = 0;
        while (previousChar < this.text.length()) {
            int nextChar = Math.min(this.text.length(), previousChar + safeStride);
            if (nextChar < this.text.length()
                    && Character.isLowSurrogate(this.text.charAt(nextChar))
                    && nextChar > previousChar) {
                nextChar--;
            }
            if (nextChar <= previousChar) {
                nextChar = Math.min(this.text.length(), previousChar + 1);
            }
            byteOffset += encodedLength(previousChar, nextChar);
            chars.add(nextChar);
            bytes.add(byteOffset);
            previousChar = nextChar;
        }

        charAnchors = toIntArray(chars);
        byteAnchors = toIntArray(bytes);
    }

    int byteOffsetForCharIndex(int charIndex) {
        int safeIndex = normalizeCharBoundary(Math.max(0, Math.min(charIndex, text.length())));
        int anchor = floorIndex(charAnchors, safeIndex);
        return byteAnchors[anchor] + encodedLength(charAnchors[anchor], safeIndex);
    }

    int charIndexForByteOffset(long byteOffset) {
        int target = (int) Math.max(0L, Math.min(byteOffset, totalBytes()));
        int anchor = floorIndex(byteAnchors, target);
        int charIndex = charAnchors[anchor];
        int currentByte = byteAnchors[anchor];
        while (charIndex < text.length()) {
            int nextChar = charIndex + Character.charCount(text.codePointAt(charIndex));
            int nextByte = currentByte + encodedLength(charIndex, nextChar);
            if (nextByte > target) break;
            charIndex = nextChar;
            currentByte = nextByte;
        }
        return charIndex;
    }

    int totalBytes() {
        return byteAnchors[byteAnchors.length - 1];
    }

    private int normalizeCharBoundary(int index) {
        if (index > 0 && index < text.length() && Character.isLowSurrogate(text.charAt(index))) {
            return index - 1;
        }
        return index;
    }

    private int encodedLength(int start, int end) {
        if (end <= start) return 0;
        return text.substring(start, end).getBytes(charset).length;
    }

    private static int floorIndex(int[] values, int target) {
        int low = 0;
        int high = values.length - 1;
        while (low <= high) {
            int mid = (low + high) >>> 1;
            if (values[mid] <= target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return Math.max(0, high);
    }

    private static int[] toIntArray(List<Integer> values) {
        int[] result = new int[values.size()];
        for (int i = 0; i < values.size(); i++) result[i] = values.get(i);
        return result;
    }
}
